### ISSUE TITLE
Allow multi-statement macros

### ISSUE THREAD

**Issue by sfackler** [rust-lang/rust#10681]
Macros can now expand to multiple items, but not multiple statements. This is a bit sad, especially since a macro that can expand in a module can't expand in a function. This will probably require a bit of restructuring in the parser, since there doesn't seem to be a good way to parse a series of expressions. `parser::parse_block_tail` is almost right, but would need to be pulled apart a bit.

**Comment by alexcrichton** [comment]
@sfackler I feel like this issue is still open, but the original example you have compiles today. Do you have an updated example for this?

**Comment by sfackler** [comment]
@alexcrichton 

```rust
macro_rules! foo (
    () => (
        struct Foo;
        struct Bar;
    )
);

fn main() {
    foo!();
}
```

**Comment by durka** [comment]
The situation is improved by #24155 but there are still hygiene issues ([reported](https://github.com/rust-lang/rust/pull/24155#issuecomment-126487021)). Here is an example: http://is.gd/sGjll7

**Comment by marjakm** [comment]
Another example: http://is.gd/oZTig7
Also broken if I give in an ident for vector: http://is.gd/rg1Hh1
The workaround in the durkas example doesn't work in this case because of the lifetime. I couldn't find a way to work around this issue.

**Comment by durka** [comment]
triage: multi statement macros are still allowed, but broken, because hygiene runs separately on each statement, so any bindings introduced are not accessible, even within the macro expansion.

**Comment by jseyfried** [comment]
This was fixed in #32923.

**Comment by alexcrichton** [comment]
Yay!
